{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "c90f8355-8f6c-4378-949b-7f90db0b1016",
   "metadata": {},
   "outputs": [],
   "source": [
    "# !pip install torch torchvision torchaudio --extra-index-url https://download.pytorch.org/whl/cu123"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "01cacbfc-bb6d-4a47-8944-82230b5e94a3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "device(type='cuda')"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# using Kernel PyTorch-1.10\n",
    "# using Kernel PyTorch-1.10\n",
    "# using Kernel PyTorch-1.10\n",
    "\n",
    "from gpn.data import GenomeMSA #, Tokenizer\n",
    "import gpn.model\n",
    "\n",
    "# from tqdm import tqdm\n",
    "from sklearn.preprocessing import LabelEncoder, StandardScaler\n",
    "from transformers import AutoModel #, AutoModelForMaskedLM\n",
    "import torch\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "# import seaborn as sns\n",
    "# import matplotlib.pyplot as plt\n",
    "# import pickle\n",
    "import re\n",
    "import os\n",
    "import csv\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore')\n",
    "\n",
    "%run preprocess_utility.py\n",
    "\n",
    "device=torch.device('cuda' if torch.cuda.is_available() else 'cpu')\n",
    "device"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "a662182d-8566-4312-8f56-3a1cbb72dfb3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading MSA...\n",
      "Loading MSA... Done\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Some weights of the model checkpoint at songlab/gpn-msa-sapiens were not used when initializing GPNRoFormerModel: ['cls.predictions.transform.LayerNorm.bias', 'cls.predictions.transform.dense.bias', 'cls.predictions.decoder.bias', 'cls.predictions.bias', 'cls.predictions.transform.dense.weight', 'cls.predictions.transform.LayerNorm.weight', 'cls.predictions.decoder.weight']\n",
      "- This IS expected if you are initializing GPNRoFormerModel from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).\n",
      "- This IS NOT expected if you are initializing GPNRoFormerModel from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).\n"
     ]
    }
   ],
   "source": [
    "model_path = \"songlab/gpn-msa-sapiens\"\n",
    "# msa_path = \"zip:///::89.zarr.zip\"\n",
    "msa_path = \"zip:///::/home/sunhuaikuan/ondemand/blue_gpn/examples/msa/89.zarr.zip\"\n",
    "genome_msa = GenomeMSA(msa_path)\n",
    "model = AutoModel.from_pretrained(model_path).to(device)\n",
    "model.eval();"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "61f1f950-2739-44b5-b7c0-f666d3c80d6a",
   "metadata": {},
   "source": [
    "### Main Function to get Embedding"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "2c8d7807-85ad-44b3-9447-63e55064a87f",
   "metadata": {},
   "outputs": [],
   "source": [
    "comp = {'A':1, 'C':2, 'G':3, 'T':4}\n",
    "\n",
    "max_seqlen=128\n",
    "\n",
    "def Genosome2Embedding(chrom, pos_start, pos_end, y):  #  ref,alt,\n",
    "    # tokenized msa\n",
    "    # pos_start = pos - 65\n",
    "    # pos_end=pos_start + 128\n",
    "    msa = genome_msa.get_msa(str(chrom), pos_start, pos_end, strand=\"+\", tokenize=True)\n",
    "    # print(msa.shape)\n",
    "\n",
    "    msa = torch.tensor(np.expand_dims(msa, 0).astype(np.int64))\n",
    "    # msa\n",
    "\n",
    "    # separating human from rest of species\n",
    "    input_ids, aux_features = msa[:, :, 0], msa[:, :, 1:]\n",
    "    \n",
    "    input_ids = input_ids.to(device)\n",
    "    aux_features = aux_features.to(device)\n",
    "\n",
    "\n",
    "    with torch.no_grad():\n",
    "        last_hidden_state = model(input_ids=input_ids, aux_features=aux_features).last_hidden_state\n",
    "        \n",
    "        # Mean Pooling: Compute the mean across the sequence length (dim=1)\n",
    "        mean_pooled = last_hidden_state.mean(dim=1)  # Shape: (batch_size, embedding_dim)\n",
    "\n",
    "    feature=np.append(mean_pooled.cpu().numpy(),  [ y])  # chrom,  pos_end-pos_start, comp[ref],comp[alt],\n",
    "    \n",
    "    return feature\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "477b3f25-8134-4403-840b-1ee587329800",
   "metadata": {},
   "source": [
    "### Output CSV File"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "64c65f65-0068-44a6-aacb-efb384aabe38",
   "metadata": {},
   "outputs": [],
   "source": [
    "def output2CSV(df, csv_Filename):\n",
    "\n",
    "    if os.path.exists(csv_Filename):\n",
    "        os.remove(csv_Filename)\n",
    "\n",
    "    rows=[]\n",
    "    # progress_bar = tqdm(total=df.shape[0], desc=\"Processing\")\n",
    "    for index, row in df.iterrows():\n",
    "        \n",
    "        chrom=row['CHROM']\n",
    "        pos_start=row['START']\n",
    "        pos_end=row['END']\n",
    "        # ref=row['REF']\n",
    "        # alt=row['ALT']        \n",
    "        # info=row['INFO']\n",
    "        y=row['y']\n",
    "        try:\n",
    "            embedding  =  Genosome2Embedding(chrom, pos_start,pos_end,y) # ref,alt, \n",
    "            rows.append(embedding)\n",
    "    \n",
    "        except Exception as e:\n",
    "            print(f\"exception caught: {e}\"+str(row['CHROM'])+'-'+str(row['START']))\n",
    "    \n",
    "        # progress_bar.update(1)\n",
    "\n",
    "\n",
    "        if ((index % 5000) ==0):\n",
    "            with open(csv_Filename, mode='a', newline='') as file:\n",
    "                writer = csv.writer(file)\n",
    "                # writer.writerow(rows)\n",
    "                for row in rows:\n",
    "                    writer.writerow(row)\n",
    "            rows=[]\n",
    "            # progress_bar.update(1)\n",
    "            print(f\"complete index={index}\")\n",
    "\n",
    "    \n",
    "    with open(csv_Filename, mode='a', newline='') as file:\n",
    "        writer = csv.writer(file)\n",
    "        # writer.writerow(rows)\n",
    "        for row in rows:\n",
    "            writer.writerow(row)\n",
    "\n",
    "    print(f\"Create File: \"+csv_Filename)\n",
    "    # progress_bar.reset()\n",
    "    # progress_bar.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "d9aac259-f7c8-4e0a-ac9b-d4c5e1fed293",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "### PreProcess Dataset\n",
    "# def PreProcess(data_filename):\n",
    "#     # df=pd.read_csv(data_filename, delimiter='\\t',  names=['CHROM', 'POS', 'REF','ALT','INFO'])  # No head column row\n",
    "#     df=pd.read_csv(data_filename, header=0, sep='\\t', index_col=None)   # With head column row\n",
    "\n",
    "#     df=df[(df['REF'].str.len() == 1) & (df['ALT'].str.len() == 1)]\n",
    "#     df['CHROM'] = df['CHROM'].str.replace('chr', '')\n",
    "#     df['y']=0\n",
    "#     return df\n",
    "\n",
    "# df1=PreProcess('gnomad.v4.1.exon.txt')\n",
    "# df1\n",
    "# output2CSV(df1,'./gnomad.v4.1.exon.gpn.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "14719867-8daf-47b6-8957-4f78881c9b1a",
   "metadata": {},
   "source": [
    "### Load mathylation data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "e649ff17-8b0c-422a-a2d2-e120b36d5e76",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>y</th>\n",
       "      <th>END</th>\n",
       "      <th>START</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>100</td>\n",
       "      <td>788470</td>\n",
       "      <td>788342</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>80</td>\n",
       "      <td>788535</td>\n",
       "      <td>788407</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>95</td>\n",
       "      <td>788888</td>\n",
       "      <td>788760</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>90</td>\n",
       "      <td>788907</td>\n",
       "      <td>788779</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>100</td>\n",
       "      <td>788922</td>\n",
       "      <td>788794</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54789</th>\n",
       "      <td>Y</td>\n",
       "      <td>65</td>\n",
       "      <td>56870966</td>\n",
       "      <td>56870838</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54790</th>\n",
       "      <td>Y</td>\n",
       "      <td>17</td>\n",
       "      <td>56871046</td>\n",
       "      <td>56870918</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54791</th>\n",
       "      <td>Y</td>\n",
       "      <td>34</td>\n",
       "      <td>56871172</td>\n",
       "      <td>56871044</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54792</th>\n",
       "      <td>Y</td>\n",
       "      <td>0</td>\n",
       "      <td>56871565</td>\n",
       "      <td>56871437</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54793</th>\n",
       "      <td>Y</td>\n",
       "      <td>36</td>\n",
       "      <td>56873962</td>\n",
       "      <td>56873834</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>47953 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      CHROM    y       END     START\n",
       "0         1  100    788470    788342\n",
       "1         1   80    788535    788407\n",
       "2         1   95    788888    788760\n",
       "3         1   90    788907    788779\n",
       "4         1  100    788922    788794\n",
       "...     ...  ...       ...       ...\n",
       "54789     Y   65  56870966  56870838\n",
       "54790     Y   17  56871046  56870918\n",
       "54791     Y   34  56871172  56871044\n",
       "54792     Y    0  56871565  56871437\n",
       "54793     Y   36  56873962  56873834\n",
       "\n",
       "[47953 rows x 4 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "datafile='methylation'\n",
    "\n",
    "import gzip\n",
    "import pandas as pd\n",
    "\n",
    "max_length= 128 # 186\n",
    "\n",
    "\n",
    "# Open the gzipped file and read it into a DataFrame\n",
    "data_filename = '/home/sunhuaikuan/ondemand/blue_papers/DNA_LLM_REVIEW/datasets/task05-methylation/GSM6637962_CpG_coverage20_GRCh38.bed.gz'  \n",
    "with gzip.open(data_filename, 'rt') as f:\n",
    "    df = pd.read_csv(f, sep='\\t')  # Automatically detects header from the file\n",
    "\n",
    "\n",
    "df['CHROM'] = df['CHROM'].str.replace('chr', '', regex=False)\n",
    "\n",
    "df['START']=df['FROM']- max_length //2 -1\n",
    "\n",
    "df['END']=df['START'] + max_length\n",
    "\n",
    "\n",
    "# df['REF']='A'\n",
    "# df['ALT']='A'\n",
    "# df['SIZE']=max_length\n",
    "\n",
    "df = df.rename(columns={'Percentage':'y'})\n",
    "df = df.drop(['FROM','TO','Coverage'], axis=1)\n",
    "\n",
    "\n",
    "cols = df.columns.tolist()\n",
    "\n",
    "# Move the 3rd column to the last position\n",
    "cols.append(cols.pop(2))\n",
    "\n",
    "# Reorder the DataFrame\n",
    "df = df[cols]\n",
    "df=df[~df['CHROM'].str.contains('KI',na=False)]\n",
    "df=df[~df['CHROM'].str.contains('GL',na=False)]\n",
    "df=df[~df['CHROM'].str.contains('M',na=False)]\n",
    "df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "d0117e7a-74f6-4f4e-a846-35d4f1c7fdd2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# df['CHROM'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "8de9326f-4bb1-4d5a-8490-064f30d689d9",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "# max_seqlen=128\n",
    "\n",
    "\n",
    "# # datafile='Homo_sapiens.GRCh38.109'\n",
    "# # datafile='clinvar_20240805.10missense'\n",
    "# # datafile='clinvar_20240805.missense_matched'\n",
    "# datafile='clinvar_20240805.noncoding'\n",
    "\n",
    "\n",
    "# df=pd.read_csv(datafile+'.txt', delimiter='\\t')\n",
    "\n",
    "# columns_to_keep=['CHROM','POS','ID','REF','ALT','Pathogenicity']\n",
    "# df = df[columns_to_keep]\n",
    "\n",
    "# # Merge CHROM=9 and '9' etc\n",
    "# for i in range(1,23):\n",
    "#     df.loc[df['CHROM']==i,'CHROM']=str(i)\n",
    "# # df['CHROM'].value_counts()\n",
    "\n",
    "# df['START']=df['POS']- max_seqlen //2 -1\n",
    "\n",
    "# df['END']=df['START'] + max_seqlen\n",
    "\n",
    "# df=df[~df['CHROM'].isna()]\n",
    "\n",
    "# Pathogenicity_dict={'B':0,'P':1}\n",
    "# df['y'] = df['Pathogenicity'].map(Pathogenicity_dict)\n",
    "# # df['CHROM'].value_counts()\n",
    "# df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "c26ed760-7a6c-4caf-a16b-bad70f3fb76b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "complete index=0\n",
      "complete index=5000\n",
      "complete index=10000\n",
      "complete index=15000\n",
      "complete index=20000\n",
      "complete index=25000\n",
      "complete index=30000\n",
      "complete index=35000\n",
      "complete index=40000\n",
      "complete index=45000\n",
      "Create File: methylation_gpn_embedding.csv\n",
      "CPU times: user 9min 18s, sys: 16.7 s, total: 9min 35s\n",
      "Wall time: 14min 12s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "\n",
    "output2CSV(df,datafile+'_gpn_embedding.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b20ae207-fd2a-40ec-9c03-a88b14504d6c",
   "metadata": {},
   "source": [
    "### Load CSV File"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "50eb2e37-e973-43e7-b7ea-429c62e1b9a0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>10</th>\n",
       "      <th>...</th>\n",
       "      <th>760</th>\n",
       "      <th>761</th>\n",
       "      <th>762</th>\n",
       "      <th>763</th>\n",
       "      <th>764</th>\n",
       "      <th>765</th>\n",
       "      <th>766</th>\n",
       "      <th>767</th>\n",
       "      <th>768</th>\n",
       "      <th>y</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.612933</td>\n",
       "      <td>0.057651</td>\n",
       "      <td>-1.246180</td>\n",
       "      <td>0.277122</td>\n",
       "      <td>-0.913016</td>\n",
       "      <td>-0.190393</td>\n",
       "      <td>-0.607857</td>\n",
       "      <td>0.931990</td>\n",
       "      <td>0.123141</td>\n",
       "      <td>-0.208136</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.309597</td>\n",
       "      <td>0.027291</td>\n",
       "      <td>0.728741</td>\n",
       "      <td>0.025583</td>\n",
       "      <td>-0.226388</td>\n",
       "      <td>-0.122658</td>\n",
       "      <td>-0.830180</td>\n",
       "      <td>0.287627</td>\n",
       "      <td>0.697681</td>\n",
       "      <td>80.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.439794</td>\n",
       "      <td>-0.217588</td>\n",
       "      <td>-1.365778</td>\n",
       "      <td>0.898174</td>\n",
       "      <td>-0.855323</td>\n",
       "      <td>-0.057502</td>\n",
       "      <td>-0.541308</td>\n",
       "      <td>0.825290</td>\n",
       "      <td>-0.152932</td>\n",
       "      <td>0.057166</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.478570</td>\n",
       "      <td>-0.206125</td>\n",
       "      <td>0.665346</td>\n",
       "      <td>0.037774</td>\n",
       "      <td>-0.350043</td>\n",
       "      <td>-0.394456</td>\n",
       "      <td>-0.936075</td>\n",
       "      <td>0.136769</td>\n",
       "      <td>0.755584</td>\n",
       "      <td>95.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.509560</td>\n",
       "      <td>-0.267472</td>\n",
       "      <td>-1.300921</td>\n",
       "      <td>0.996084</td>\n",
       "      <td>-0.885385</td>\n",
       "      <td>0.017741</td>\n",
       "      <td>-0.389727</td>\n",
       "      <td>0.736749</td>\n",
       "      <td>-0.024453</td>\n",
       "      <td>-0.015932</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.499621</td>\n",
       "      <td>-0.270688</td>\n",
       "      <td>0.674775</td>\n",
       "      <td>0.030219</td>\n",
       "      <td>-0.439443</td>\n",
       "      <td>-0.394929</td>\n",
       "      <td>-0.962380</td>\n",
       "      <td>0.199156</td>\n",
       "      <td>0.799231</td>\n",
       "      <td>90.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.521640</td>\n",
       "      <td>-0.312444</td>\n",
       "      <td>-1.231837</td>\n",
       "      <td>0.999817</td>\n",
       "      <td>-0.902015</td>\n",
       "      <td>0.040940</td>\n",
       "      <td>-0.388363</td>\n",
       "      <td>0.678714</td>\n",
       "      <td>0.115234</td>\n",
       "      <td>-0.002658</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.482253</td>\n",
       "      <td>-0.241800</td>\n",
       "      <td>0.587185</td>\n",
       "      <td>-0.061250</td>\n",
       "      <td>-0.514301</td>\n",
       "      <td>-0.447875</td>\n",
       "      <td>-0.878306</td>\n",
       "      <td>0.179583</td>\n",
       "      <td>0.790959</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.524937</td>\n",
       "      <td>-0.312998</td>\n",
       "      <td>-1.194757</td>\n",
       "      <td>1.013375</td>\n",
       "      <td>-0.899023</td>\n",
       "      <td>0.025699</td>\n",
       "      <td>-0.360354</td>\n",
       "      <td>0.680852</td>\n",
       "      <td>0.182384</td>\n",
       "      <td>-0.016140</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.480661</td>\n",
       "      <td>-0.231768</td>\n",
       "      <td>0.515274</td>\n",
       "      <td>-0.073867</td>\n",
       "      <td>-0.556997</td>\n",
       "      <td>-0.451583</td>\n",
       "      <td>-0.834344</td>\n",
       "      <td>0.167359</td>\n",
       "      <td>0.820079</td>\n",
       "      <td>96.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47947</th>\n",
       "      <td>1.011808</td>\n",
       "      <td>-0.340859</td>\n",
       "      <td>-0.550905</td>\n",
       "      <td>0.540617</td>\n",
       "      <td>-0.895378</td>\n",
       "      <td>-0.361927</td>\n",
       "      <td>-0.386061</td>\n",
       "      <td>0.108030</td>\n",
       "      <td>0.178881</td>\n",
       "      <td>-0.575671</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.276941</td>\n",
       "      <td>-0.365299</td>\n",
       "      <td>1.221117</td>\n",
       "      <td>0.361901</td>\n",
       "      <td>-0.237325</td>\n",
       "      <td>-0.175215</td>\n",
       "      <td>-1.330128</td>\n",
       "      <td>0.236117</td>\n",
       "      <td>0.664278</td>\n",
       "      <td>65.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47948</th>\n",
       "      <td>0.920000</td>\n",
       "      <td>-0.485343</td>\n",
       "      <td>-0.189774</td>\n",
       "      <td>0.895718</td>\n",
       "      <td>-0.681114</td>\n",
       "      <td>-0.384050</td>\n",
       "      <td>-0.271887</td>\n",
       "      <td>-0.290705</td>\n",
       "      <td>0.370207</td>\n",
       "      <td>-0.471858</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.058086</td>\n",
       "      <td>-0.546873</td>\n",
       "      <td>0.958808</td>\n",
       "      <td>0.289077</td>\n",
       "      <td>-0.237743</td>\n",
       "      <td>-0.465524</td>\n",
       "      <td>-1.134478</td>\n",
       "      <td>0.210529</td>\n",
       "      <td>0.553010</td>\n",
       "      <td>17.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47949</th>\n",
       "      <td>0.988282</td>\n",
       "      <td>-0.167276</td>\n",
       "      <td>-0.147423</td>\n",
       "      <td>-0.331529</td>\n",
       "      <td>-0.631022</td>\n",
       "      <td>-0.981502</td>\n",
       "      <td>-0.874661</td>\n",
       "      <td>0.072641</td>\n",
       "      <td>0.650392</td>\n",
       "      <td>-0.628087</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.723373</td>\n",
       "      <td>-0.058664</td>\n",
       "      <td>0.731405</td>\n",
       "      <td>0.305108</td>\n",
       "      <td>0.059723</td>\n",
       "      <td>0.031155</td>\n",
       "      <td>-1.170336</td>\n",
       "      <td>-0.064444</td>\n",
       "      <td>0.422516</td>\n",
       "      <td>34.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47950</th>\n",
       "      <td>0.748591</td>\n",
       "      <td>-0.345141</td>\n",
       "      <td>-0.674358</td>\n",
       "      <td>0.946595</td>\n",
       "      <td>-0.722178</td>\n",
       "      <td>-0.245178</td>\n",
       "      <td>-0.144482</td>\n",
       "      <td>0.193256</td>\n",
       "      <td>0.020578</td>\n",
       "      <td>-0.512533</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.154925</td>\n",
       "      <td>-0.550619</td>\n",
       "      <td>0.952435</td>\n",
       "      <td>0.460786</td>\n",
       "      <td>-0.215765</td>\n",
       "      <td>-0.452896</td>\n",
       "      <td>-1.396009</td>\n",
       "      <td>0.284754</td>\n",
       "      <td>0.886341</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47951</th>\n",
       "      <td>0.544165</td>\n",
       "      <td>-0.475644</td>\n",
       "      <td>-0.755631</td>\n",
       "      <td>0.796180</td>\n",
       "      <td>-0.834340</td>\n",
       "      <td>-0.287577</td>\n",
       "      <td>-0.117117</td>\n",
       "      <td>0.228021</td>\n",
       "      <td>-0.036780</td>\n",
       "      <td>-0.558169</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.228218</td>\n",
       "      <td>-0.409220</td>\n",
       "      <td>0.773816</td>\n",
       "      <td>0.414297</td>\n",
       "      <td>-0.427338</td>\n",
       "      <td>-0.443502</td>\n",
       "      <td>-1.254795</td>\n",
       "      <td>0.404900</td>\n",
       "      <td>0.833885</td>\n",
       "      <td>36.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>47952 rows × 769 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              1         2         3         4         5         6         7  \\\n",
       "0      0.612933  0.057651 -1.246180  0.277122 -0.913016 -0.190393 -0.607857   \n",
       "1      0.439794 -0.217588 -1.365778  0.898174 -0.855323 -0.057502 -0.541308   \n",
       "2      0.509560 -0.267472 -1.300921  0.996084 -0.885385  0.017741 -0.389727   \n",
       "3      0.521640 -0.312444 -1.231837  0.999817 -0.902015  0.040940 -0.388363   \n",
       "4      0.524937 -0.312998 -1.194757  1.013375 -0.899023  0.025699 -0.360354   \n",
       "...         ...       ...       ...       ...       ...       ...       ...   \n",
       "47947  1.011808 -0.340859 -0.550905  0.540617 -0.895378 -0.361927 -0.386061   \n",
       "47948  0.920000 -0.485343 -0.189774  0.895718 -0.681114 -0.384050 -0.271887   \n",
       "47949  0.988282 -0.167276 -0.147423 -0.331529 -0.631022 -0.981502 -0.874661   \n",
       "47950  0.748591 -0.345141 -0.674358  0.946595 -0.722178 -0.245178 -0.144482   \n",
       "47951  0.544165 -0.475644 -0.755631  0.796180 -0.834340 -0.287577 -0.117117   \n",
       "\n",
       "              8         9        10  ...       760       761       762  \\\n",
       "0      0.931990  0.123141 -0.208136  ... -1.309597  0.027291  0.728741   \n",
       "1      0.825290 -0.152932  0.057166  ... -1.478570 -0.206125  0.665346   \n",
       "2      0.736749 -0.024453 -0.015932  ... -1.499621 -0.270688  0.674775   \n",
       "3      0.678714  0.115234 -0.002658  ... -1.482253 -0.241800  0.587185   \n",
       "4      0.680852  0.182384 -0.016140  ... -1.480661 -0.231768  0.515274   \n",
       "...         ...       ...       ...  ...       ...       ...       ...   \n",
       "47947  0.108030  0.178881 -0.575671  ... -1.276941 -0.365299  1.221117   \n",
       "47948 -0.290705  0.370207 -0.471858  ... -1.058086 -0.546873  0.958808   \n",
       "47949  0.072641  0.650392 -0.628087  ... -0.723373 -0.058664  0.731405   \n",
       "47950  0.193256  0.020578 -0.512533  ... -1.154925 -0.550619  0.952435   \n",
       "47951  0.228021 -0.036780 -0.558169  ... -1.228218 -0.409220  0.773816   \n",
       "\n",
       "            763       764       765       766       767       768      y  \n",
       "0      0.025583 -0.226388 -0.122658 -0.830180  0.287627  0.697681   80.0  \n",
       "1      0.037774 -0.350043 -0.394456 -0.936075  0.136769  0.755584   95.0  \n",
       "2      0.030219 -0.439443 -0.394929 -0.962380  0.199156  0.799231   90.0  \n",
       "3     -0.061250 -0.514301 -0.447875 -0.878306  0.179583  0.790959  100.0  \n",
       "4     -0.073867 -0.556997 -0.451583 -0.834344  0.167359  0.820079   96.0  \n",
       "...         ...       ...       ...       ...       ...       ...    ...  \n",
       "47947  0.361901 -0.237325 -0.175215 -1.330128  0.236117  0.664278   65.0  \n",
       "47948  0.289077 -0.237743 -0.465524 -1.134478  0.210529  0.553010   17.0  \n",
       "47949  0.305108  0.059723  0.031155 -1.170336 -0.064444  0.422516   34.0  \n",
       "47950  0.460786 -0.215765 -0.452896 -1.396009  0.284754  0.886341    0.0  \n",
       "47951  0.414297 -0.427338 -0.443502 -1.254795  0.404900  0.833885   36.0  \n",
       "\n",
       "[47952 rows x 769 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# def load_embedding_file(csv_filename):\n",
    "\n",
    "#     df=pd.read_csv(csv_filename)\n",
    "    \n",
    "#     column_names = [f'{i}' for i in range(1, df.shape[1]-2)]\n",
    "#     column_names.extend(['CHROM', 'SIZE',  'y'])  # 'REF', 'ALT', \n",
    "    \n",
    "#     df.columns = column_names\n",
    "#     return df\n",
    "\n",
    "df = load_embedding_file(datafile+'_gpn_embedding.csv')\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6e2bb830-9748-48ac-9ca7-7d591ae52e2a",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "PyTorch-1.10",
   "language": "python",
   "name": "pytorch-1.10"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
