{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "3c6e319c-468a-49f1-bfbf-ed11118721dc",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/blue/xiaofan/sunhuaikuan/anaconda3/envs/dnabert2/lib/python3.8/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html\n",
      "  from .autonotebook import tqdm as notebook_tqdm\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "device(type='cuda')"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import torch\n",
    "from transformers import AutoTokenizer, AutoModel, BertConfig\n",
    "from transformers.models.bert.configuration_bert import BertConfig\n",
    "\n",
    "import os\n",
    "import csv\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from tqdm import tqdm\n",
    "\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore')\n",
    "\n",
    "\n",
    "%run preprocess_utility.py\n",
    "\n",
    "device=torch.device('cuda' if torch.cuda.is_available() else 'cpu')\n",
    "device"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "d18b4f0c-2254-4557-abcd-6a785e54b964",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>START</th>\n",
       "      <th>END</th>\n",
       "      <th>TYPE</th>\n",
       "      <th>CLUSTER</th>\n",
       "      <th>SIZE</th>\n",
       "      <th>ROWID</th>\n",
       "      <th>y</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>7</td>\n",
       "      <td>116953541</td>\n",
       "      <td>116953691</td>\n",
       "      <td>first_exon</td>\n",
       "      <td>first_exon</td>\n",
       "      <td>150</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>12</td>\n",
       "      <td>54241755</td>\n",
       "      <td>54241905</td>\n",
       "      <td>first_exon</td>\n",
       "      <td>first_exon</td>\n",
       "      <td>150</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>20</td>\n",
       "      <td>38033948</td>\n",
       "      <td>38034098</td>\n",
       "      <td>first_exon</td>\n",
       "      <td>first_exon</td>\n",
       "      <td>150</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>15</td>\n",
       "      <td>82659560</td>\n",
       "      <td>82659710</td>\n",
       "      <td>first_exon</td>\n",
       "      <td>first_exon</td>\n",
       "      <td>150</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>9</td>\n",
       "      <td>131579497</td>\n",
       "      <td>131579647</td>\n",
       "      <td>first_exon</td>\n",
       "      <td>first_exon</td>\n",
       "      <td>150</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33713</th>\n",
       "      <td>9</td>\n",
       "      <td>124692442</td>\n",
       "      <td>124692551</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>smallRNA</td>\n",
       "      <td>109</td>\n",
       "      <td>33713</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33714</th>\n",
       "      <td>9</td>\n",
       "      <td>128244721</td>\n",
       "      <td>128244830</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>smallRNA</td>\n",
       "      <td>109</td>\n",
       "      <td>33714</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33715</th>\n",
       "      <td>X</td>\n",
       "      <td>45746157</td>\n",
       "      <td>45746266</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>smallRNA</td>\n",
       "      <td>109</td>\n",
       "      <td>33715</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33716</th>\n",
       "      <td>X</td>\n",
       "      <td>45747015</td>\n",
       "      <td>45747124</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>smallRNA</td>\n",
       "      <td>109</td>\n",
       "      <td>33716</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33717</th>\n",
       "      <td>X</td>\n",
       "      <td>66018870</td>\n",
       "      <td>66018979</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>smallRNA</td>\n",
       "      <td>109</td>\n",
       "      <td>33717</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>33508 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      CHROM      START        END        TYPE     CLUSTER  SIZE  ROWID  y\n",
       "0         7  116953541  116953691  first_exon  first_exon   150      0  0\n",
       "1        12   54241755   54241905  first_exon  first_exon   150      1  0\n",
       "2        20   38033948   38034098  first_exon  first_exon   150      2  0\n",
       "3        15   82659560   82659710  first_exon  first_exon   150      3  0\n",
       "4         9  131579497  131579647  first_exon  first_exon   150      4  0\n",
       "...     ...        ...        ...         ...         ...   ...    ... ..\n",
       "33713     9  124692442  124692551       miRNA    smallRNA   109  33713  6\n",
       "33714     9  128244721  128244830       miRNA    smallRNA   109  33714  6\n",
       "33715     X   45746157   45746266       miRNA    smallRNA   109  33715  6\n",
       "33716     X   45747015   45747124       miRNA    smallRNA   109  33716  6\n",
       "33717     X   66018870   66018979       miRNA    smallRNA   109  33717  6\n",
       "\n",
       "[33508 rows x 8 columns]"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "datafile_path = '../../datasets/task03-genomic-regions/Homo_sapiens.GRCh38.109.txt.gz'  \n",
    "df = preprocess_home_sapiens_datafile(datafile_path)\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "458f1f04-cedd-492d-8ca2-9ddd943a1d07",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "pseudogene               4998\n",
       "first_intron             4985\n",
       "ncRNA_gene               4975\n",
       "first_three_prime_UTR    4971\n",
       "first_five_prime_UTR     4952\n",
       "first_exon               4926\n",
       "smallRNA                 3701\n",
       "Name: CLUSTER, dtype: int64"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['CLUSTER'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "bb084a09-419c-4d1b-b057-33f03ec390b2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "333.17406962785117"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.loc[df['CLUSTER'] == 'pseudogene', 'SIZE'].mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "bef1d132-c776-4ea2-94ce-fa68553c66a2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "247.46258776328986"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.loc[df['CLUSTER'] == 'first_intron', 'SIZE'].mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "59c4770e-9d35-4ab4-8eb0-368a7754d914",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "388.17608040201003"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.loc[df['CLUSTER'] == 'ncRNA_gene', 'SIZE'].mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "d67952b6-6dd7-4428-b89b-d8aadbed4251",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "324.2804264735466"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.loc[df['CLUSTER'] == 'first_three_prime_UTR', 'SIZE'].mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "4d3ec07b-4dac-4029-83bf-d7dce914923a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "235.85238287560583"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.loc[df['CLUSTER'] == 'first_five_prime_UTR', 'SIZE'].mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "53859a95-c655-4a09-8f4a-2c03278b700d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "227.30024360535933"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.loc[df['CLUSTER'] == 'first_exon', 'SIZE'].mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "a634ccc5-e6f0-47e1-b613-6b7d019485b6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "86.66414482572277"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.loc[df['CLUSTER'] == 'smallRNA', 'SIZE'].mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "d0d72473-16b0-4526-acaf-8d6f79b49774",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Some weights of the model checkpoint at zhihan1996/DNABERT-2-117M were not used when initializing BertModel: ['cls.predictions.transform.dense.weight', 'cls.predictions.decoder.weight', 'cls.predictions.decoder.bias', 'cls.predictions.transform.dense.bias', 'cls.predictions.transform.LayerNorm.weight', 'cls.predictions.transform.LayerNorm.bias']\n",
      "- This IS expected if you are initializing BertModel from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).\n",
      "- This IS NOT expected if you are initializing BertModel from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).\n",
      "Some weights of BertModel were not initialized from the model checkpoint at zhihan1996/DNABERT-2-117M and are newly initialized: ['bert.pooler.dense.bias', 'bert.pooler.dense.weight']\n",
      "You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.\n"
     ]
    }
   ],
   "source": [
    "\n",
    "tokenizer = AutoTokenizer.from_pretrained(\"zhihan1996/DNABERT-2-117M\", trust_remote_code=True)\n",
    "config = BertConfig.from_pretrained(\"zhihan1996/DNABERT-2-117M\")\n",
    "model = AutoModel.from_pretrained(\"zhihan1996/DNABERT-2-117M\", trust_remote_code=True, config=config).to(device)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0495a5df-3a69-4196-b12d-ab504d00f4d4",
   "metadata": {},
   "source": [
    "### Load Human Chrom Sequences from .fa File"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "f5760522-3c27-4a0c-9fe0-209a212edc62",
   "metadata": {},
   "outputs": [],
   "source": [
    "fasta_file = \"../genome.hg38rg.fa\"\n",
    "chrom_sequences = read_fasta(fasta_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "105ff3f0-4ba8-4920-bc33-7a96c0656ba5",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_embedding(dna):\n",
    "\n",
    "    inputs = tokenizer(dna, return_tensors = 'pt')[\"input_ids\"].to(device)\n",
    "\n",
    "    with torch.no_grad():\n",
    "        hidden_states = model(inputs)[0]  # Get model outputs\n",
    "        last_hidden_state = hidden_states[0]  # Access the last hidden state directly\n",
    "\n",
    "    # Extract the [CLS] embedding\n",
    "    cls_embedding = last_hidden_state[0, :]  # Extract the [CLS] embedding\n",
    "\n",
    "    return cls_embedding"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "86d7cc90-10f4-4b8f-9b44-48cfadc4222b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_subsequence(chrom_name, start_pos, length):\n",
    "    \n",
    "    # print(chrom_name, start_pos, length)\n",
    "    if chrom_name in chrom_sequences:\n",
    "        sequence = chrom_sequences[chrom_name]\n",
    "        subsequence = sequence[start_pos:start_pos + length]\n",
    "        return subsequence\n",
    "    else:\n",
    "        raise ValueError(f\"Chromosome '{chrom_name}' not found in the FASTA file.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "9bc056b3-d504-495e-852e-2851ad0d2bb6",
   "metadata": {},
   "outputs": [],
   "source": [
    "def append_rows_to_csv(csv_Filename, rows):\n",
    "    with open(csv_Filename, mode='a', newline='') as file:\n",
    "        writer = csv.writer(file)\n",
    "        for row in rows:\n",
    "            writer.writerow(row)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d4c1ea2f-6b84-4308-a4d0-4e018b1d446f",
   "metadata": {},
   "source": [
    "### Main Process"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "91c20946-3db1-4bf0-bb7a-28903c76ceee",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "index = 0 completed\n",
      "index = 1000 completed\n",
      "index = 2000 completed\n",
      "index = 3000 completed\n",
      "index = 4000 completed\n",
      "index = 5000 completed\n",
      "index = 6000 completed\n",
      "index = 7000 completed\n",
      "index = 8000 completed\n",
      "index = 9000 completed\n",
      "index = 10000 completed\n",
      "index = 11000 completed\n",
      "index = 12000 completed\n",
      "index = 13000 completed\n",
      "index = 14000 completed\n",
      "index = 15000 completed\n",
      "index = 16000 completed\n",
      "index = 17000 completed\n",
      "index = 18000 completed\n",
      "index = 19000 completed\n",
      "index = 20000 completed\n",
      "index = 21000 completed\n",
      "index = 22000 completed\n",
      "index = 23000 completed\n",
      "index = 24000 completed\n",
      "index = 25000 completed\n",
      "index = 26000 completed\n",
      "index = 27000 completed\n",
      "index = 28000 completed\n",
      "index = 29000 completed\n",
      "index = 30000 completed\n",
      "index = 31000 completed\n",
      "index = 32000 completed\n",
      "index = 33000 completed\n",
      "Create File: ./homo_sapiens_dnabert2_embedding.csv\n",
      "CPU times: user 5min 29s, sys: 1.19 s, total: 5min 30s\n",
      "Wall time: 6min 7s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "\n",
    "comp = {'A':1, 'C':2, 'G':3, 'T':4}\n",
    "\n",
    "max_length=512\n",
    "\n",
    "csv_Filename = './homo_sapiens_dnabert2_embedding.csv'\n",
    "if os.path.exists(csv_Filename):\n",
    "    os.remove(csv_Filename)\n",
    "\n",
    "\n",
    "datafile_path = '../../datasets/task03-homo-sapiens/Homo_sapiens.GRCh38.109.txt.gz'  \n",
    "df = preprocess_home_sapiens_datafile(datafile_path)\n",
    "\n",
    "df.loc[df['SIZE'] > max_length, 'END'] = df['START'] + max_length\n",
    "df.loc[df['SIZE'] > max_length, 'SIZE'] = max_length\n",
    "df.drop(columns=['END','TYPE','CLUSTER'], inplace=True)\n",
    "# df\n",
    "\n",
    "rows=[]\n",
    "for index, row in df.iterrows():      \n",
    "    chrom=row['CHROM']\n",
    "    pos_start=row['START']\n",
    "    if pos_start<=1:\n",
    "        pos_start=1\n",
    "    rowid=row['ROWID']\n",
    "    y=row['y']\n",
    "    length = row['SIZE']\n",
    "    \n",
    "    subsequence = get_subsequence(chrom, pos_start, length)\n",
    "    if 'N' in subsequence:\n",
    "        print(\"The character 'N' is present in the string.\")\n",
    "        \n",
    "    embedding = get_embedding(subsequence)\n",
    "    # print(embedding.shape)\n",
    "\n",
    "    # feature=np.array(embedding_df.iloc[64])\n",
    "    rows.append(np.append(embedding.cpu().detach().numpy(),  [ rowid, y]))\n",
    "\n",
    "    if ((index % 2000) ==0):\n",
    "        append_rows_to_csv(csv_Filename, rows)\n",
    "        rows=[]\n",
    "        print (f\"index = {index} completed\")\n",
    "        \n",
    "append_rows_to_csv(csv_Filename, rows)\n",
    "\n",
    "print(f\"Create File: \"+csv_Filename)\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e93f1c14-0ef8-4efa-84df-c94596ac0b59",
   "metadata": {},
   "source": [
    "### Load CSV File"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "90d6b579-726c-438b-9c2b-1c91b2e81b6a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>10</th>\n",
       "      <th>...</th>\n",
       "      <th>761</th>\n",
       "      <th>762</th>\n",
       "      <th>763</th>\n",
       "      <th>764</th>\n",
       "      <th>765</th>\n",
       "      <th>766</th>\n",
       "      <th>767</th>\n",
       "      <th>768</th>\n",
       "      <th>ROWID</th>\n",
       "      <th>y</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>-0.052749</td>\n",
       "      <td>0.097021</td>\n",
       "      <td>-0.040993</td>\n",
       "      <td>-0.346654</td>\n",
       "      <td>-0.228739</td>\n",
       "      <td>-0.028974</td>\n",
       "      <td>0.314350</td>\n",
       "      <td>-0.232598</td>\n",
       "      <td>-0.286229</td>\n",
       "      <td>-0.153878</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.122888</td>\n",
       "      <td>0.098309</td>\n",
       "      <td>-0.241941</td>\n",
       "      <td>0.123506</td>\n",
       "      <td>0.112383</td>\n",
       "      <td>0.345726</td>\n",
       "      <td>-0.009172</td>\n",
       "      <td>-0.232611</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.041690</td>\n",
       "      <td>-0.149883</td>\n",
       "      <td>0.112252</td>\n",
       "      <td>-0.104266</td>\n",
       "      <td>0.043270</td>\n",
       "      <td>-0.174862</td>\n",
       "      <td>-0.019622</td>\n",
       "      <td>0.083466</td>\n",
       "      <td>0.092115</td>\n",
       "      <td>0.129391</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.084495</td>\n",
       "      <td>-0.035526</td>\n",
       "      <td>-0.103098</td>\n",
       "      <td>-0.096645</td>\n",
       "      <td>0.282093</td>\n",
       "      <td>0.071169</td>\n",
       "      <td>0.225908</td>\n",
       "      <td>-0.232176</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>-0.223088</td>\n",
       "      <td>-0.252698</td>\n",
       "      <td>0.238166</td>\n",
       "      <td>-0.163469</td>\n",
       "      <td>-0.071502</td>\n",
       "      <td>0.305217</td>\n",
       "      <td>0.130825</td>\n",
       "      <td>-0.125792</td>\n",
       "      <td>0.002430</td>\n",
       "      <td>-0.027796</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.222114</td>\n",
       "      <td>0.230668</td>\n",
       "      <td>-0.297031</td>\n",
       "      <td>-0.200689</td>\n",
       "      <td>-0.019629</td>\n",
       "      <td>0.158987</td>\n",
       "      <td>0.125832</td>\n",
       "      <td>0.101306</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>-0.246262</td>\n",
       "      <td>-0.045285</td>\n",
       "      <td>-0.143605</td>\n",
       "      <td>-0.057505</td>\n",
       "      <td>-0.053923</td>\n",
       "      <td>0.136253</td>\n",
       "      <td>-0.029290</td>\n",
       "      <td>-0.291272</td>\n",
       "      <td>-0.286636</td>\n",
       "      <td>-0.224322</td>\n",
       "      <td>...</td>\n",
       "      <td>0.013754</td>\n",
       "      <td>-0.042559</td>\n",
       "      <td>-0.304382</td>\n",
       "      <td>-0.128556</td>\n",
       "      <td>-0.014183</td>\n",
       "      <td>-0.054658</td>\n",
       "      <td>-0.004198</td>\n",
       "      <td>-0.140027</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>-0.239890</td>\n",
       "      <td>0.128498</td>\n",
       "      <td>0.239283</td>\n",
       "      <td>0.033655</td>\n",
       "      <td>-0.266111</td>\n",
       "      <td>-0.010146</td>\n",
       "      <td>-0.093944</td>\n",
       "      <td>0.070097</td>\n",
       "      <td>-0.157701</td>\n",
       "      <td>0.294032</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.085098</td>\n",
       "      <td>0.080050</td>\n",
       "      <td>-0.031489</td>\n",
       "      <td>-0.122065</td>\n",
       "      <td>0.021016</td>\n",
       "      <td>0.050609</td>\n",
       "      <td>0.141866</td>\n",
       "      <td>0.200009</td>\n",
       "      <td>5.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33502</th>\n",
       "      <td>0.017066</td>\n",
       "      <td>-0.111009</td>\n",
       "      <td>0.022180</td>\n",
       "      <td>-0.261551</td>\n",
       "      <td>-0.271820</td>\n",
       "      <td>-0.162913</td>\n",
       "      <td>0.190877</td>\n",
       "      <td>-0.169575</td>\n",
       "      <td>0.006689</td>\n",
       "      <td>0.179100</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.030896</td>\n",
       "      <td>-0.036130</td>\n",
       "      <td>-0.062570</td>\n",
       "      <td>-0.123316</td>\n",
       "      <td>0.205648</td>\n",
       "      <td>0.112838</td>\n",
       "      <td>0.123704</td>\n",
       "      <td>0.236961</td>\n",
       "      <td>33713.0</td>\n",
       "      <td>6.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33503</th>\n",
       "      <td>-0.121790</td>\n",
       "      <td>-0.037089</td>\n",
       "      <td>0.006528</td>\n",
       "      <td>-0.118168</td>\n",
       "      <td>-0.280798</td>\n",
       "      <td>-0.099464</td>\n",
       "      <td>0.028415</td>\n",
       "      <td>0.005333</td>\n",
       "      <td>-0.223928</td>\n",
       "      <td>-0.348742</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.287694</td>\n",
       "      <td>0.021488</td>\n",
       "      <td>-0.167536</td>\n",
       "      <td>0.144424</td>\n",
       "      <td>0.048377</td>\n",
       "      <td>0.103088</td>\n",
       "      <td>0.146799</td>\n",
       "      <td>0.023316</td>\n",
       "      <td>33714.0</td>\n",
       "      <td>6.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33504</th>\n",
       "      <td>-0.048092</td>\n",
       "      <td>-0.305211</td>\n",
       "      <td>-0.069179</td>\n",
       "      <td>-0.200773</td>\n",
       "      <td>-0.300104</td>\n",
       "      <td>-0.013643</td>\n",
       "      <td>0.447924</td>\n",
       "      <td>-0.122317</td>\n",
       "      <td>-0.286923</td>\n",
       "      <td>-0.055943</td>\n",
       "      <td>...</td>\n",
       "      <td>0.035547</td>\n",
       "      <td>-0.012060</td>\n",
       "      <td>-0.154126</td>\n",
       "      <td>0.243896</td>\n",
       "      <td>0.189965</td>\n",
       "      <td>0.304383</td>\n",
       "      <td>0.006802</td>\n",
       "      <td>-0.142035</td>\n",
       "      <td>33715.0</td>\n",
       "      <td>6.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33505</th>\n",
       "      <td>-0.184854</td>\n",
       "      <td>-0.242769</td>\n",
       "      <td>0.011729</td>\n",
       "      <td>-0.293413</td>\n",
       "      <td>-0.274668</td>\n",
       "      <td>-0.121231</td>\n",
       "      <td>0.203303</td>\n",
       "      <td>-0.101680</td>\n",
       "      <td>0.109956</td>\n",
       "      <td>0.164296</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.041231</td>\n",
       "      <td>-0.229076</td>\n",
       "      <td>-0.249888</td>\n",
       "      <td>-0.023293</td>\n",
       "      <td>0.185503</td>\n",
       "      <td>0.099200</td>\n",
       "      <td>-0.165828</td>\n",
       "      <td>-0.165640</td>\n",
       "      <td>33716.0</td>\n",
       "      <td>6.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33506</th>\n",
       "      <td>-0.289235</td>\n",
       "      <td>0.144573</td>\n",
       "      <td>-0.282267</td>\n",
       "      <td>-0.082688</td>\n",
       "      <td>-0.273460</td>\n",
       "      <td>-0.080090</td>\n",
       "      <td>0.277865</td>\n",
       "      <td>-0.210119</td>\n",
       "      <td>-0.462484</td>\n",
       "      <td>0.067385</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.094622</td>\n",
       "      <td>0.133447</td>\n",
       "      <td>0.005300</td>\n",
       "      <td>-0.179174</td>\n",
       "      <td>0.296637</td>\n",
       "      <td>0.113235</td>\n",
       "      <td>0.003118</td>\n",
       "      <td>-0.233746</td>\n",
       "      <td>33717.0</td>\n",
       "      <td>6.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>33507 rows × 770 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              1         2         3         4         5         6         7  \\\n",
       "0     -0.052749  0.097021 -0.040993 -0.346654 -0.228739 -0.028974  0.314350   \n",
       "1      0.041690 -0.149883  0.112252 -0.104266  0.043270 -0.174862 -0.019622   \n",
       "2     -0.223088 -0.252698  0.238166 -0.163469 -0.071502  0.305217  0.130825   \n",
       "3     -0.246262 -0.045285 -0.143605 -0.057505 -0.053923  0.136253 -0.029290   \n",
       "4     -0.239890  0.128498  0.239283  0.033655 -0.266111 -0.010146 -0.093944   \n",
       "...         ...       ...       ...       ...       ...       ...       ...   \n",
       "33502  0.017066 -0.111009  0.022180 -0.261551 -0.271820 -0.162913  0.190877   \n",
       "33503 -0.121790 -0.037089  0.006528 -0.118168 -0.280798 -0.099464  0.028415   \n",
       "33504 -0.048092 -0.305211 -0.069179 -0.200773 -0.300104 -0.013643  0.447924   \n",
       "33505 -0.184854 -0.242769  0.011729 -0.293413 -0.274668 -0.121231  0.203303   \n",
       "33506 -0.289235  0.144573 -0.282267 -0.082688 -0.273460 -0.080090  0.277865   \n",
       "\n",
       "              8         9        10  ...       761       762       763  \\\n",
       "0     -0.232598 -0.286229 -0.153878  ... -0.122888  0.098309 -0.241941   \n",
       "1      0.083466  0.092115  0.129391  ... -0.084495 -0.035526 -0.103098   \n",
       "2     -0.125792  0.002430 -0.027796  ... -0.222114  0.230668 -0.297031   \n",
       "3     -0.291272 -0.286636 -0.224322  ...  0.013754 -0.042559 -0.304382   \n",
       "4      0.070097 -0.157701  0.294032  ... -0.085098  0.080050 -0.031489   \n",
       "...         ...       ...       ...  ...       ...       ...       ...   \n",
       "33502 -0.169575  0.006689  0.179100  ... -0.030896 -0.036130 -0.062570   \n",
       "33503  0.005333 -0.223928 -0.348742  ... -0.287694  0.021488 -0.167536   \n",
       "33504 -0.122317 -0.286923 -0.055943  ...  0.035547 -0.012060 -0.154126   \n",
       "33505 -0.101680  0.109956  0.164296  ... -0.041231 -0.229076 -0.249888   \n",
       "33506 -0.210119 -0.462484  0.067385  ... -0.094622  0.133447  0.005300   \n",
       "\n",
       "            764       765       766       767       768    ROWID    y  \n",
       "0      0.123506  0.112383  0.345726 -0.009172 -0.232611      1.0  0.0  \n",
       "1     -0.096645  0.282093  0.071169  0.225908 -0.232176      2.0  0.0  \n",
       "2     -0.200689 -0.019629  0.158987  0.125832  0.101306      3.0  0.0  \n",
       "3     -0.128556 -0.014183 -0.054658 -0.004198 -0.140027      4.0  0.0  \n",
       "4     -0.122065  0.021016  0.050609  0.141866  0.200009      5.0  0.0  \n",
       "...         ...       ...       ...       ...       ...      ...  ...  \n",
       "33502 -0.123316  0.205648  0.112838  0.123704  0.236961  33713.0  6.0  \n",
       "33503  0.144424  0.048377  0.103088  0.146799  0.023316  33714.0  6.0  \n",
       "33504  0.243896  0.189965  0.304383  0.006802 -0.142035  33715.0  6.0  \n",
       "33505 -0.023293  0.185503  0.099200 -0.165828 -0.165640  33716.0  6.0  \n",
       "33506 -0.179174  0.296637  0.113235  0.003118 -0.233746  33717.0  6.0  \n",
       "\n",
       "[33507 rows x 770 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "def load_embedding_file(csv_filename):\n",
    "\n",
    "    df=pd.read_csv(csv_filename)\n",
    "        \n",
    "    column_names = [f'{i}' for i in range(1, df.shape[1]-1)]\n",
    "    column_names.extend(['ROWID',  'y']) \n",
    "    \n",
    "    df.columns = column_names\n",
    "    return df\n",
    "\n",
    "df = load_embedding_file('./homo_sapiens_dnabert2_embedding.csv')\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e67bc911-88b5-4cf7-87a3-73f86a930540",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "DNABert2",
   "language": "python",
   "name": "dnabert2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.20"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
